storage/innobase/dict/dict0dict.cc

```diff
@@ -7450,27 +7450,85 @@ dberr_t dict_get_dictionary_info_by_id(ulint dict_id, char **name,
   return err;
 }
 
-bool dict_detect_encryption(bool is_upgrade, space_id_t mysql_plugin_space) {
-  bool encrypt_mysql = false;
-  if (is_upgrade) {
-    if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
-      return (srv_sys_space.is_encrypted());
-    }
+static bool dict_should_mysql_ibd_be_created_encrypted_due_to_upgrade(
+    space_id_t mysql_plugin_space) {
+  if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
+    return (srv_sys_space.is_encrypted());
+  }
 
     space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
     if (space_id == SPACE_UNKNOWN) {
       return (false);
     }
 
-    fil_space_t *space = fil_space_get(space_id);
-    ut_ad(space != nullptr);
+  const fil_space_t *space = fil_space_get(space_id);
+  ut_ad(space != nullptr);
 
-    if (space == nullptr) {
-      return (false);
-    }
-    encrypt_mysql = FSP_FLAGS_GET_ENCRYPTION(space->flags);
+  if (space == nullptr) {
+    return false;
+  }
+  return FSP_FLAGS_GET_ENCRYPTION(space->flags);
+}
+
+static bool dict_should_mysql_ibd_be_created_encrypted() {
+  return srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE;
+}
+
+static bool dict_should_mysql_ibd_be_opened_as_encrypted(const char *filepath,
+                                                         bool &success) {
+  pfs_os_file_t file = os_file_create_simple_no_error_handling(
+      innodb_data_file_key, filepath, OS_FILE_OPEN, OS_FILE_READ_ONLY,
+      srv_read_only_mode, &success);
+
+  if (!success) {
+    return (false);
   }
 
-  return (encrypt_mysql || srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
-          srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE);
+  /* Read the first page of the tablespace */
+  std::unique_ptr<byte, std::function<void(byte *)>> buf(
```

Comment:
This looks useful enough to have it in e.g. ut0ut.h: unique_ptr with ut_free deleter (but with struct ut_free_deleter w/ operator() instead of std::function<>/lambda to avoid having state in the smart pointer)

Reply:
added to ut0new.h if this is OK